flang/lib/Optimizer/Builder/FIRBuilder.cpp

```diff
@@ -872,10 +963,13 @@ void fir::factory::genRecordAssignment(fir::FirOpBuilder &builder,
   }
   // Otherwise, the derived type has compile time constant size and for which
   // the component by component assignment can be replaced by a memory copy.
-  auto rhsVal = fir::getBase(rhs);
-  if (fir::isa_ref_type(rhsVal.getType()))
-    rhsVal = builder.create<fir::LoadOp>(loc, rhsVal);
-  builder.create<fir::StoreOp>(loc, rhsVal, fir::getBase(lhs));
+  // Since we do not know the size of the derived type in lowering, do a
+  // component by component assignment. Note that a single fir.load/fir.store
+  // could be used on "small" record types, but as the type size grows, this
+  // leads to issues in LLVM (long compile times, long IR files, and even
+  // asserts at some point). Since there is no good size boundary, just always
+  // use component by component assignment here.
+  genComponentByComponentAssignment(builder, loc, lhs, rhs);
```

Comment:
Eventually someone might want to investigate that boundary in more detail, and use the more performant code in some cases.  Would it be useful to leave a copy of the old code somewhere in a comment or under some form of `if (false)` condition to help with that?  A developer option is another alternative, although that might be overkill.

Reply:
It is not clear load/store is more performant in any cases. It is textually a lot more appealing/simple, but LLVM anyway transform it into an unrolled component by component load/stores at the assembly level (by unrolled I mean that if you have a component that is an array, it won't loop over it, but create a load/store for each and every element).
But I agree that it would be interesting to measure this at some point. I think the code doing full load/store was so trivial that there is no point keeping it and having to justify why we are keeping it when upstreaming.